File: gym-tracker-frontend/src/types/exercise.ts
export type ExerciseID = string;
export type EquipmentID = string;
export type EquipmentType = 'dumbbell' | 'barbell' | 'machine';

export interface Exercise {
    id: ExerciseID;
    name: string;
    description: string;
    primary_patterns: PatternContribution[];
    secondary_patterns: PatternContribution[];
    derived_muscles: {
        primary: string[];
        secondary: string[];
    };
    equipment: EquipmentType[];
    source: 'system' | 'user';
    is_custom: boolean;
}

export interface PatternContribution {
    pattern: string;
    pattern_name: string;
    contribution: number;
    range_of_motion: string;
    notes: string;
}

export interface Equipment {
    id: EquipmentID;
    type: EquipmentType;
    name: string;
    manufacturer?: string;
    model_number?: string;
}

File: internal/domain/exercise.go
package domain

import (
	"errors"
	"time"
)

type ExerciseID string

type Exercise struct {
	ID          ExerciseID
	Name        string
	Description string
	UserID      string

	// Muscles are derived from patterns automatically
	PrimaryPatterns    []MovementPatternContribution
	SecondaryPatterns  []MovementPatternContribution
	EquipmentUsed      Equipment
	SuggestedEquipment []EquipmentType

	CreatedAt time.Time
	UpdatedAt time.Time
}

func (e *Exercise) Validate() error {
	if e.Name == "" {
		return errors.New("exercise name is required")
	}
	return nil
}

type MovementPatternContribution struct {
	Pattern MovementPattern

	// 0.0 - 1.0, how much this pattern is emphasized
	Contribution float64

	// "full", "partial", "lengthened", "shortened"
	RangeOfMotion string

	Notes string
}
